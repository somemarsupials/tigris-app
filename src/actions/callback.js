import axios from 'axios';
import env from '../env';
import { prefix } from '../lib';

export const ACTIONS = prefix('CALLBACK', [
  'SUCCESS',
  'FAILURE',
]);

function success(provider, data, store = localStorage) {
  return function thunk(dispatch) {
    store.setItem(provider, JSON.stringify(data));
    return dispatch({
      type: ACTIONS.SUCCESS,
      data: data,
    });
  };
}

function failure(error) {
  return {
    type: ACTIONS.FAILURE,
    error: error,
  };
}

const OUTCOMES = {
  success: success,
  failure: failure,
};

function getToken(provider, httpClient = axios, outcomes = OUTCOMES) {
  return async function thunk(dispatch) {
    let response;

    try {
      response = await httpClient({
        method: 'get',
        url: `${env.api}/auth/${provider}/token`,
        withCredentials: true,
      });
      return dispatch(outcomes.success(provider, response.data));
    }
    catch (error) {
      let data = error.response || {
        status: 500,
        data: {
          error: error.message,
        },
      };
      return dispatch(outcomes.failure(data));
    }
  };
}

export default {
  success,
  failure,
  getToken,
};
