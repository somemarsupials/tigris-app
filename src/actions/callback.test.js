import actions from './callback';

describe('callback actions', () => {
  let dispatch;

  beforeEach(() => {
    dispatch = jest.fn();
  });

  describe('#success', () => {
    let store;

    beforeEach(() => {
      store = { setItem: jest.fn() };
      actions.success('spotify', { a: 1 }, store)(dispatch);
    });

    it('calls setItem with stringified data', () => {
      expect(store.setItem).toHaveBeenCalledWith('spotify', '{"a":1}');
    });

    it('dispatches action', () => {
      expect(dispatch.mock.calls).toMatchSnapshot();
    });
  });

  describe('#failure', () => {
    beforeEach(() => {
      actions.failure({ error: 'problem' })
    });

    it('dispatches action', () => {
      expect(dispatch.mock.calls).toMatchSnapshot();
    });
  });

  describe('#getToken', () => {
    let httpClient, outcomes;

    beforeEach(() => {
      httpClient = jest.fn();
      outcomes = {
        success: jest.fn().mockReturnValue('y'),
        failure: jest.fn().mockReturnValue('n'),
      };
    });

    describe('when non-http error', () => {
      let error;

      beforeEach(() => {
        httpClient.mockImplementation(() => {
          error = new Error();
          throw error;
        });
        actions.getToken('spotify', httpClient, outcomes)(dispatch);
      });

      it('calls failure', () => {
        expect(outcomes.failure).toHaveBeenCalledWith(error);
      });

      it('dispatches action', () => {
        expect(dispatch).toHaveBeenCalledWith('n');
      });
    });

    describe('when error', () => {
      let error;

      beforeEach(() => {
        httpClient.mockImplementation(() => {
          error = new Error();
          error.response = {
            status: 404,
            message: 'message',
          };
          throw error;
        });
        actions.getToken('spotify', httpClient, outcomes)(dispatch);
      });

      it('calls failure', () => {
        expect(outcomes.failure).toHaveBeenCalledWith(error.response);
      });
      it('dispatches action', () => {
        expect(dispatch).toHaveBeenCalledWith('n');
      });
    });

    describe('when success', () => {
      beforeEach(() => {
        httpClient.mockReturnValue({ data: [1, 2, 3] });
        actions.getToken('spotify', httpClient, outcomes)(dispatch);
      });

      it('calls success', () => {
        expect(outcomes.success).toHaveBeenCalledWith('spotify', [1, 2, 3]);
      });

      it('dispatches action', () => {
        expect(dispatch).toHaveBeenCalledWith('y');
      });
    });
  });
});
