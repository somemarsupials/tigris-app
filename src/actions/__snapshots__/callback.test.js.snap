// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callback actions #failure dispatches action 1`] = `Array []`;

exports[`callback actions #success dispatches action 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "a": 1,
      },
      "type": "CALLBACK/SUCCESS",
    },
  ],
]
`;
